lib/surrealist/builder.rb

```diff
@@ -87,30 +87,68 @@ def take_values_from_instance(instance:, value:, hash:, key:, schema:, method:)
           parse_hash(hash: value, schema: hash, instance: result, key: key)
         else
           type = value
-          assign_value(method: key, value: result, type: type) do |coerced_value|
+          assign_value(instance: instance, method: key, value: result, type: type) do |coerced_value|
             schema[method] = schema[method].merge(key => coerced_value)
           end
         end
       end
 
       # Assigns value returned from a method to a corresponding key in the schema hash.
       #
+      # @param [Object] instance the instance of the object which methods from the schema are called on.
       # @param [Symbol] method a key from the schema hash representing a method on the instance.
       # @param [Object] value a value that has to be type-checked.
       # @param [Class] type class representing data type.
       #
       # @raise +Surrealist::InvalidTypeError+ if type-check failed at some point.
       #
       # @return [Hash] schema
-      def assign_value(method:, value:, type:, &_block)
+      def assign_value(instance:, method:, value:, type:, &_block)
         if TypeHelper.valid_type?(value: value, type: type)
           value = TypeHelper.coerce(type: type, value: value)
-          yield value
+
+          @stack ||= []
+
+          if value.respond_to?(:build_schema)
+            yield assign_nested_record(instance: instance, value: value)
+          elsif value.respond_to?(:each) && value.any? { |v| Helper.surrealist?(v.class) }
+            yield assign_nested_collection(instance: instance, value: value)
+          else
+            yield value
+          end
         else
           raise Surrealist::InvalidTypeError,
                 "Wrong type for key `#{method}`. Expected #{type}, got #{value.class}."
         end
       end
+
+      # Assists in recursively generating schema for records while preventing infinite self-referencing
+      #
+      # @param [Object] instance the instance of the object which methods from the schema are called on.
+      # @param [Object] value a value that has to be type-checked.
+      #
+      # @return [Array] of schemas
+      def assign_nested_collection(instance:, value:)
+        return if @stack.include?(value.first.class)
+        @stack << instance.class << value.first.class
```

Comment:
I don't really get why you have decided to push `instance.class` here if it will be deleted after one line, could you please expand on this? Or it is like a recursion termination condition?

Reply:
Yes, this tracks each class that we are diving into to build the schema and prevents us from having infinite recursion